man/man1/ndb.1
.TH NDB 1
.SH NAME
ndbquery, ndbmkhash, ndbmkdb, ndbipquery, ndbmkhosts \- network database
.SH SYNOPSIS
.B ndbquery
[
.B -f
.I dbfile
]
.I "attr value"
[
.I rattr
]
.br
.B ndbipquery
.I "attr value"
.I rattr...
.br
.B ndbmkhash
.I "file attr"
.br
.B ndbmkdb
.SH DESCRIPTION
The network database holds administrative information used by
.I authdial
(see
.MR authsrv 3 )
and
.MR secstored 1 .
.PP
.I Ndbquery 
searches the database for an attribute of type
.I attr
and value
.IR value .
If
.I rattr
is not specified, all entries matched by the search are returned.
If 
.I rattr
is specified, the value of the first pair with attribute
.I rattr
of all the matched entries is returned.
.PP
.I Ndbipquery
uses
.I ndbipinfo
(see
.MR ndb 3 )
to search for the values of the attributes
.I rattr
corresponding to the system
with entries of attribute type
.I attr
and
value
.IR value .
.PP
.I Ndbmkhash
creates a hash file for all entries with attribute
.I attr
in database file
.IR file .
The hash files are used by 
.I ndbquery
and by the ndb library routines.
.\" .PP
.\" .I Ndb/cs
.\" is a server used by
.\" .IR dial (2)
.\" to translate network names.
.\" It is started at boot time.
.\" It finds out what networks are configured
.\" by looking for
.\" .B /net/*/clone
.\" when it starts.
.\" It can also be told about networks by writing
.\" to
.\" .B /net/cs
.\" a message of the form:
.\" .IP
.\" .B "add net1 net2 ..."
.\" .PP
.\" .I Ndb/cs
.\" also sets the system name in
.\" .B /dev/sysname
.\" if it can figure it out.
.\" The options are:
.\" .TP
.\" .B -f
.\" supplies the name of the data base file to use,
.\" default
.\" .BR /lib/ndb/local .
.\" .TP
.\" .B -x
.\" specifies the mount point of the
.\" network.
.\" .TP
.\" .B -n
.\" causes cs to do nothing but set the system name.
.\" .PP
.\" .I Ndb/csquery
.\" can be used to query
.\" .I ndb/cs
.\" to see how it resolves addresses.
.\" .I Ndb/csquery
.\" prompts for addresses and prints out what
.\" .I ndb/cs
.\" returns.
.\" .I Server
.\" defaults to
.\" .BR /net/cs .
.\" If any
.\" .I addrs
.\" are specified,
.\" .I ndb/csquery
.\" prints their translations and immediately exits.
.\" The exit status will be nil only if all addresses 
.\" were successfully translated
.\" The
.\" .B -s
.\" flag sets exit status without printing any results.
.\" .PP
.\" .I Ndb/dns
.\" is a server used by
.\" .I ndb/cs
.\" and by remote systems to translate Internet domain names.
.\" .I Ndb/dns
.\" is started at boot time.
.\" By default
.\" .I dns
.\" serves only requests written to
.\" .BR /net/dns .
.\" The options are:
.\" .TP
.\" .B -f
.\" supplies the name of the data base file to use,
.\" default
.\" .BR /lib/ndb/local .
.\" .TP
.\" .B -x
.\" specifies the mount point of the
.\" network.
.\" .TP
.\" .B -s
.\" also answer domain requests sent to UDP port 53.
.\" .TP
.\" .B -n
.\" whenever a zone that we serve changes, send UDP NOTIFY
.\" messages to any dns slaves for that zone.
.\" .TP
.\" .B -z
.\" whenever we receive a UDP NOTIFY message, run
.\" .I program
.\" with the domain name of the area as its argument.
.\" .TP
.\" .B -r
.\" defer to other servers to resolve queries.
.\" .PP
.\" When the
.\" .B -r
.\" option is specified, the servers used come from the
.\" .I dns
.\" attribute in the database.  For example, to specify a set of dns servers that
.\" will resolve requests for systems on the network
.\" .IR mh-net :
.\" .EX
.\" 
.\" ipnet=mh-net ip=135.104.0.0 ipmask=255.255.0.0
.\" 	dns=ns1.cs.bell-labs.com
.\" 	dns=ns2.cs.bell-labs.com
.\" dom=ns1.cs.bell-labs.com ip=135.104.1.11
.\" dom=ns2.cs.bell-labs.com ip=135.104.1.12
.\" 
.\" .EE
.\" .PP
.\" The server for a domain is indicated by a database entry containing
.\" both a
.\" .I dom
.\" and a
.\" .I ns
.\" attribute.
.\" For example, the entry for the Internet root is:
.\" .EX
.\" 
.\" dom=
.\" 	ns=A.ROOT-SERVERS.NET
.\" 	ns=B.ROOT-SERVERS.NET
.\" 	ns=C.ROOT-SERVERS.NET
.\" dom=A.ROOT-SERVERS.NET ip=198.41.0.4
.\" dom=B.ROOT-SERVERS.NET ip=128.9.0.107
.\" dom=C.ROOT-SERVERS.NET ip=192.33.4.12
.\" 
.\" .EE
.\" The last three lines provide a mapping for the
.\" server names to their ip addresses.  This is only
.\" a hint and will be superseded from whatever is learned
.\" from servers owning the domain.
.\" .PP
.\" You can also serve a subtree of the domain name space from the local
.\" database.  You indicate subtrees that you'ld like to serve by
.\" adding an
.\" .B soa=
.\" attribute to the root entry.
.\" For example, the Bell Labs CS research domain is:
.\" .EX
.\" 
.\" dom=cs.bell-labs.com soa=
.\" 	refresh=3600 ttl=3600
.\" 	ns=plan9.bell-labs.com
.\" 	ns=ns1.cs.bell-labs.com
.\" 	ns=ns2.cs.bell-labs.com
.\" 	mb=presotto@plan9.bell-labs.com
.\" 	mx=mail.research.bell-labs.com pref=20
.\" 	mx=plan9.bell-labs.com pref=10
.\" 	dnsslave=nslocum.cs.bell-labs.com
.\" 	dnsslave=vex.cs.bell-labs.com
.\" 
.\" .EE
.\" Here, the
.\" .B mb
.\" entry is the mail address of the person responsible for the
.\" domain (default
.\" .BR postmaster ).
.\" The
.\" .B mx
.\" entries list mail exchangers for the domain name and
.\" .B refresh
.\" and
.\" .B ttl
.\" define the area refresh interval and the minimum TTL for
.\" records in this domain.
.\" The
.\" .B dnsslave
.\" entries specify slave DNS servers that should be notified
.\" when the domain changes.  The notification also requires
.\" the
.\" .B -n
.\" flag.
.\" .PP
.\" You can also serve reverse lookups (returning the name that
.\" goes with an IP address) by adding an
.\" .B soa=
.\" attribute to the entry defining the root of the reverse space.
.\" For example, to provide reverse lookup for all addresses in
.\" starting with 135.104 you must have a record like:
.\" .EX
.\" 
.\" dom=104.135.in-addr.arpa soa=
.\" 	refresh=3600 ttl=3600
.\" 	ns=plan9.bell-labs.com
.\" 	ns=ns1.cs.bell-labs.com
.\" 	ns=ns2.cs.bell-labs.com
.\" .EE
.\" Notice the form of the reverse address, i.e., it's the bytes of the
.\" address range you are serving reversed and with
.\" .B .in-addr.arpa
.\" appended.  This is a standard form for a domain name in an IPv4 PTR record.
.\" .PP
.\" If such an entry exists in the database, reverse addresses will
.\" automaticly be generated from any IP addresses in the database
.\" that are under this root.  For example
.\" .EX
.\" 
.\" dom=ns1.cs.bell-labs.com ip=135.104.1.11
.\" .EE
.\" will automaticly create both forward and reverse entries for
.\" .B ns1.cs.bell-labs.com .
.\" Unlike other DNS servers, there's no way to generate
.\" inconsistent forward and reverse entries.
.\" .PP
.\" Delegation of a further subtree to another set of name servers
.\" is indicated by an
.\" .B soa=delegated
.\" attribute.
.\" .EX
.\" 
.\" dom=bignose.cs.research.bell-labs.com
.\" 	soa=delegated
.\" 	ns=anna.cs.research.bell-labs.com
.\" 	ns=dj.cs.research.bell-labs.com
.\" 
.\" .EE
.\" Nameservers within the delegated domain (as in this example)
.\" must have their IP addresses listed elsewhere in
.\" .I ndb
.\" files.
.\" .PP
.\" Wild-carded domain names can also be used.
.\" For example, to specify a mail forwarder for all Bell Labs research systems:
.\" .EX
.\" 
.\" dom=*.research.bell-labs.com
.\" 	mx=research.bell-labs.com
.\" 
.\" .EE
.\" `Cname' aliases may be established by adding a
.\" .B cname
.\" attribute giving the real domain name;
.\" the name attached to the
.\" .B dom
.\" attribute is the alias.
.\" `Cname' aliases are severely restricted;
.\" the aliases may have no other attributes than
.\" .B dom
.\" and are daily further restricted in their use by new RFCs.
.\" .EX
.\" 
.\" cname=anna.cs.research.bell-labs.com dom=www.cs.research.bell-labs.com
.\" 
.\" .EE
.\" .I Ndb/dnsquery
.\" can be used to query
.\" .I ndb/dns
.\" to see how it resolves requests.
.\" .I Ndb/dnsquery
.\" prompts for commands of the form
.\" .IP
.\" .I "domain-name request-type"
.\" .LP
.\" where
.\" .I request-type
.\" can be
.\" .BR ip ,
.\" .BR mx ,
.\" .BR ns ,
.\" .BR cname ,
.\" .BR ptr ....
.\" In the case of the inverse query type,
.\" .BR ptr ,
.\" .I dnsquery
.\" will reverse the ip address and tack on the
.\" .B .in-addr.arpa
.\" for you.
.\" .PP
.\" .I Ndb/dnsdebug
.\" is like
.\" .I ndb/dnsquery
.\" but bypasses the local server.
.\" It communicates via UDP with the domain name servers
.\" in the same way that the local resolver would and displays
.\" all packets received.
.\" The query can be specified on the command line or
.\" can be prompted for.
.\" The queries look like those of
.\" .I ndb/dnsquery
.\" with one addition.
.\" .I Ndb/dnsdebug
.\" can be directed to query a particular name server by
.\" the command
.\" .BI @ name-server\f1.
.\" From that point on, all queries go to that name server
.\" rather than being resolved by
.\" .IR dnsdebug .
.\" The
.\" .B @
.\" command returns query resolution to
.\" .IR dnsdebug .
.\" Finally, any command preceded by a
.\" .BI @ name-server
.\" sets the name server only for that command.
.\" .PP
.\" Normally
.\" .I dnsdebug
.\" uses the
.\" .B /net
.\" interface and the database file
.\" .BR /lib/ndb/local.
.\" The
.\" .B -x
.\" option directs
.\" .I dnsdebug
.\" to use the
.\" .B /net.alt
.\" interface and
.\" .B /lib/ndb/external
.\" file.
.\" The
.\" .B -r
.\" option is the same as for
.\" .IR ndb/dns .
.PP
.I Ndbmkdb
is used in concert with
.MR awk 1
scripts to convert
uucp systems files and IP host files
into database files.
It is very specific to the situation at Murray Hill.
.PP
When the database files change underfoot,
running programs
track them properly.  Nonetheless, to keep the database searches efficient
it is necessary to run
.I ndbmkhash
whenever the files are modified.
It may be profitable to control this by a frequent
.MR cron 8
job.
.PP
.I Ndbmkhosts
generates a BSD style 
.BR hosts ,
.BR hosts.txt ,
and
.B hosts.equiv
files from ndb data base files specified on the
command line (default
.B \*9/ndb/local
and
.BR \*9/ndb/friends ).
It only processes hosts whose domain names end in
.IR domname .
The output files are named
.BI db. domname \fR,
.BI equiv. domname \fR,
and
.BI txt. domname \fR.
For historical reasons, the default
.I domname
is
.BR research.att.com.
.SH EXAMPLE
.IP
.EX
% ndbquery sys helix
sys=helix dom=helix.research.bell-labs.com bootf=/mips/9powerboot
	ip=135.104.117.31 ether=080069020427
	proto=il 
.EE
.SH FILES
.TP
.B \*9/ndb/local
first database file searched
.TP
.B \*9/ndb/local.*
hash files for
.B \*9/ndb/local
.SH SOURCE
.B \*9/src/cmd/ndb
.SH SEE ALSO
.MR ndb 3 ,
.MR ndb 7


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/cmd/ip/dhcpd/ndb.c
/*
 *  this currently only works for ethernet bootp's -- presotto
 */
#include <u.h>
#include <libc.h>
#include <ip.h>
#include <bio.h>
#include <ndb.h>
#include "dat.h"

static void check72(Info *iip);

Ndb *db;
char *ndbfile;

Iplifc*
findlifc(uchar *ip)
{
	uchar x[IPaddrlen];
	Ipifc *ifc;
	Iplifc *lifc;

	for(ifc = ipifcs; ifc; ifc = ifc->next){
		for(lifc = ifc->lifc; lifc != nil; lifc = lifc->next){
			if(lifc->net[0] == 0)
				continue;
			maskip(ip, lifc->mask, x);
			if(memcmp(x, lifc->net, IPaddrlen) == 0)
				return lifc;
		}
	}
	return nil;
}

int
forme(uchar *ip)
{
	Ipifc *ifc;
	Iplifc *lifc;

extern uchar xmyipaddr[IPaddrlen];

if(memcmp(ip, xmyipaddr, IPaddrlen) == 0)
	return 1;

	for(ifc = ipifcs; ifc; ifc = ifc->next){
		for(lifc = ifc->lifc; lifc != nil; lifc = lifc->next)
			if(memcmp(ip, lifc->ip, IPaddrlen) == 0)
				return 1;
	}
	return 0;
}

uchar noetheraddr[6];

static void
setipaddr(uchar *addr, char *ip)
{
	if(ipcmp(addr, IPnoaddr) == 0)
		parseip(addr, ip);
}

static void
setipmask(uchar *mask, char *ip)
{
	if(ipcmp(mask, IPnoaddr) == 0)
		parseipmask(mask, ip);
}

/*
 *  do an ipinfo with defaults
 */
int
lookupip(uchar *ipaddr, Info *iip, int gate)
{
	char ip[32];
	Ndbtuple *t, *nt;
	char *attrs[32], **p;

	if(db == 0)
		db = ndbopen(ndbfile);
	if(db == 0){
		fprint(2, "can't open db\n");
		return -1;
	}

	p = attrs;
	*p++ = "ip";
	*p++ = "ipmask";
	*p++ = "@ipgw";
	if(!gate){
		*p++ = "bootf";
		*p++ = "bootf2";
		*p++ = "@tftp";
		*p++ = "@tftp2";
		*p++ = "rootpath";
		*p++ = "dhcp";
		*p++ = "vendorclass";
		*p++ = "ether";
		*p++ = "dom";
		*p++ = "@fs";
		*p++ = "@auth";
	}
	*p = 0;

	memset(iip, 0, sizeof(*iip));
	snprint(ip, sizeof(ip), "%I", ipaddr);
	t = ndbipinfo(db, "ip", ip, attrs, p - attrs);
	if(t == nil)
		return -1;

	for(nt = t; nt != nil; nt = nt->entry){
		if(strcmp(nt->attr, "ip") == 0)
			setipaddr(iip->ipaddr, nt->val);
		else
		if(strcmp(nt->attr, "ipmask") == 0)
			setipmask(iip->ipmask, nt->val);
		else
		if(strcmp(nt->attr, "fs") == 0)
			setipaddr(iip->fsip, nt->val);
		else
		if(strcmp(nt->attr, "auth") == 0)
			setipaddr(iip->auip, nt->val);
		else
		if(strcmp(nt->attr, "tftp") == 0)
			setipaddr(iip->tftp, nt->val);
		else
		if(strcmp(nt->attr, "tftp2") == 0)
			setipaddr(iip->tftp2, nt->val);
		else
		if(strcmp(nt->attr, "ipgw") == 0)
			setipaddr(iip->gwip, nt->val);
		else
		if(strcmp(nt->attr, "ether") == 0){
			if(memcmp(iip->etheraddr, noetheraddr, 6) == 0)
				parseether(iip->etheraddr, nt->val);
			iip->indb = 1;
		}
		else
		if(strcmp(nt->attr, "dhcp") == 0){
			if(iip->dhcpgroup[0] == 0)
				strcpy(iip->dhcpgroup, nt->val);
		}
		else
		if(strcmp(nt->attr, "bootf") == 0){
			if(iip->bootf[0] == 0)
				strcpy(iip->bootf, nt->val);
		}
		else
		if(strcmp(nt->attr, "bootf2") == 0){
			if(iip->bootf2[0] == 0)
				strcpy(iip->bootf2, nt->val);
		}
		else
		if(strcmp(nt->attr, "vendor") == 0){
			if(iip->vendor[0] == 0)
				strcpy(iip->vendor, nt->val);
		}
		else
		if(strcmp(nt->attr, "dom") == 0){
			if(iip->domain[0] == 0)
				strcpy(iip->domain, nt->val);
		}
		else
		if(strcmp(nt->attr, "rootpath") == 0){
			if(iip->rootpath[0] == 0)
				strcpy(iip->rootpath, nt->val);
		}
	}
	ndbfree(t);
	maskip(iip->ipaddr, iip->ipmask, iip->ipnet);
	return 0;
}

static uchar zeroes[6];

/*
 *  lookup info about a client in the database.  Find an address on the
 *  same net as riip.
 */
int
lookup(Bootp *bp, Info *iip, Info *riip)
{
	Ndbtuple *t, *nt;
	Ndbs s;
	char *hwattr;
	char *hwval, hwbuf[33];
	uchar ciaddr[IPaddrlen];

	if(db == 0)
		db = ndbopen(ndbfile);
	if(db == 0){
		fprint(2, "can't open db\n");
		return -1;
	}

	memset(iip, 0, sizeof(*iip));

	/* client knows its address? */
	v4tov6(ciaddr, bp->ciaddr);
	if(validip(ciaddr)){
		if(lookupip(ciaddr, iip, 0) < 0)
			return -1;	/* don't know anything about it */

check72(iip);

		if(!samenet(riip->ipaddr, iip)){
			warning(0, "%I not on %I", ciaddr, riip->ipnet);
			return -1;
		}

		/*
		 *  see if this is a masquerade, i.e., if the ether
		 *  address doesn't match what we expected it to be.
		 */
		if(memcmp(iip->etheraddr, zeroes, 6) != 0)
		if(memcmp(bp->chaddr, iip->etheraddr, 6) != 0)
			warning(0, "ciaddr %I rcvd from %E instead of %E",
				ciaddr, bp->chaddr, iip->etheraddr);

		return 0;
	}

	if(bp->hlen > Maxhwlen)
		return -1;
	switch(bp->htype){
	case 1:
		hwattr = "ether";
		hwval = hwbuf;
		snprint(hwbuf, sizeof(hwbuf), "%E", bp->chaddr);
		break;
	default:
		syslog(0, blog, "not ethernet %E, htype %d, hlen %d",
			bp->chaddr, bp->htype, bp->hlen);
		return -1;
	}

	/*
	 *  use hardware address to find an ip address on
	 *  same net as riip
	 */
	t = ndbsearch(db, &s, hwattr, hwval);
	while(t){
		for(nt = t; nt; nt = nt->entry){
			if(strcmp(nt->attr, "ip") != 0)
				continue;
			parseip(ciaddr, nt->val);
			if(lookupip(ciaddr, iip, 0) < 0)
				continue;
			if(samenet(riip->ipaddr, iip)){
				ndbfree(t);
				return 0;
			}
		}
		ndbfree(t);
		t = ndbsnext(&s, hwattr, hwval);
	}
	return -1;
}

/*
 *  interface to ndbipinfo
 */
Ndbtuple*
lookupinfo(uchar *ipaddr, char **attr, int n)
{
	char ip[32];

	sprint(ip, "%I", ipaddr);
	return ndbipinfo(db, "ip", ip, attr, n);
}

/*
 *  return the ip addresses for a type of server for system ip
 */
int
lookupserver(char *attr, uchar **ipaddrs, Ndbtuple *t)
{
	Ndbtuple *nt;
	int rv = 0;

	for(nt = t; rv < 2 && nt != nil; nt = nt->entry)
		if(strcmp(nt->attr, attr) == 0){
			parseip(ipaddrs[rv], nt->val);
			rv++;
		}
	return rv;
}

/*
 *  just lookup the name
 */
void
lookupname(char *val, Ndbtuple *t)
{
	Ndbtuple *nt;

	for(nt = t; nt != nil; nt = nt->entry)
		if(strcmp(nt->attr, "dom") == 0){
			strcpy(val, nt->val);
			break;
		}
}

uchar slash120[IPaddrlen] = { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
				0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0 };
uchar net72[IPaddrlen] = { 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
				0x0, 0x0, 0xff, 0xff, 135, 104, 72, 0 };

static void
check72(Info *iip)
{
	uchar net[IPaddrlen];

	maskip(iip->ipaddr, slash120, net);
	if(ipcmp(net, net72) == 0)
		syslog(0, blog, "check72 %I %M gw %I", iip->ipaddr, iip->ipmask, iip->gwip);
}
